big-neon/bn-api
# Create and manage discounts

Fixes: #446

This branch addresses the creation and management aspect of #446. 

Primarily:
1. CRUD related endpoints for discount creation and updating (linked to events as per the mockups)
2. Validations for discounts for data integrity where needed
3. Ability to associate ticket types for a given event with the discounts via a vector of `ticket_type_ids`
4. Associated unit, functional, and postman tests for discounts behavior

Outside of scope:
- Most of what we discussed in https://github.com/big-neon/bn-api/issues/446#issuecomment-434578054 -- I was just clarifying for many for future use cases but most of that is for the use of discount codes which is a big issue in it of itself.

Validations:
- Start before end date
- Redemption code 6 char min and confirm no other model level validations missing
- Redemption code unique on a per event level (constraint and validation) across multiple models
- Confirm ticket type is valid for discount

Future validations (when we deal with those issues / implementations) but not done here in this branch to restrict scope (making a note here so we keep track of it as considered adding them here but opted not):
- Order is within discount window during purchasing
- Ticket association with discount code to check max per user
- Max uses does not exceed discount max uses via association lookup (constraint and validation)

Not sure if we want to do the following (probably worth their own issues if so):
1. Switch holds to use discount_id and apply the discount when the hold is activated? Perhaps using the same redemption_code as the hold?
2. Switch the validation for redemption code to constrict it to just a given event and remove uniqueness constraint on redemption code per table we currently have? One downside of that approach is if multiple events have active redemption codes with the same value they'd be triggered and show discounts (although not likely to happen). Just noticed the docs mentioned to make it unique per event but right now we have it system unique.

Note:
1. Added a method to set a discount on an order but leaving the work there for pricing, etc. for another issue to constrain scope here.
2. For now only made the redemption codes unique across discounts and holds -- wasn't sure what the comp specific redemption codes shown in the mocks were used for but should be easy enough to add if those are used in the same logic.
3. Tried and failed to get diesel's dsl working with an subquery resulting in an array for `Discount::find_for_event` so ended up with plain SQL. May be worth refactoring later but I spent way too long attempting to get it to work and reading through the docs so decided it wasn't worth the effort since the plain SQL works just as well.

```
discounts::table
    .filter(discounts::event_id.eq(event_id))
    .select((
        discounts::id,
        discounts::name,
        discounts::event_id,
        discounts::redemption_code,
        discounts::max_uses,
        discounts::discount_in_cents,
        discounts::start_date,
        discounts::end_date,
        discounts::max_tickets_per_user,
        discounts::created_at,
        discounts::updated_at,
        dsl::array::<dUuid, _>(
            ticket_type_discounts::table
                .filter(ticket_type_discounts::discount_id.eq(discounts::id))
                .select(ticket_type_discounts::ticket_type_id)
        ),
    )).get_results(conn)
    .to_db_error(
        ErrorCode::QueryError,
        "Could not retrieve discounts for event",
    )
```
